pkg/cmd/cli/secrets/source.go

```diff
@@ -0,0 +1,255 @@
+package secrets
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client"
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+
+	cmdutil "github.com/openshift/origin/pkg/cmd/util"
+
+	"github.com/spf13/cobra"
+)
+
+const (
+	// CreateSourceSecretRecommendedCommandName represents name of subcommand for `oc secrets` command
+	CreateSourceSecretRecommendedCommandName = "new-sourcecfg"
+
+	createSourceSecretLong = `
+Create a new source secret
+
+Source secrets are used to authenticate against SCM servers.
+
+When creating applications, you may have a SCM server that requires authentication. In order for the
+nodes to clone source code on your behalf, they have to have the credentials. You can provide this
+information by creating a sourcecfg secret and attaching it to your service account.`
+
+	createSourceSecretExample = `  // If your source authentication method requires only username and password or token, add it by using:
+  $ %[1]s SECRET_NAME --username=USERNAME --password=PASSWORD
+
+  // If your source authentication method requires private ssh key, add it by using:
+  $ %[1]s SECRET_NAME --ssh-privatekey=FILENAME
+
+  // If you do already have a .gitconfig file needed for authentication, you can create a gitconfig secret by using:
+  $ %[2]s SECRET_NAME path/to/.gitconfig`
+
+	optionGroupExclusivityError = `must provide only one kind of option group.
+  Basic authentication group flags: --username, --password, --ca-cert
+  SSH authentication group flags: --ssh-privatekey
+`
+)
+
+// CreateSourceSecretOptions holds the credential needed to authenticate against SCM servers.
+type CreateSourceSecretOptions struct {
+	SecretNamespace string
+	SecretName      string
+	Username        string
+	Password        string
+	PrivateKeyPath  string
+	SourceURL       string
+	CertificatePath string
+	GitConfigPath   string
+
+	PromptForPassword bool
+
+	Reader io.Reader
+	Out    io.Writer
+
+	SecretsInterface client.SecretsInterface
+}
+
+const (
+	// SourceUsername is the key of the optional username for basic authentication group
+	SourceUsername = "username"
+	// SourcePassword is the key of the optional password or token for basic authentication group
+	SourcePassword = "password"
+	// SourceCertificate is the key of the optional certificate authority for basic authentication group
+	SourceCertificate = "ca.crt"
+	// SourcePrivateKey is the key of the required SSH private key for SSH authentication group
+	SourcePrivateKey = "ssh-privatekey"
+	// SourceGitconfig is the key of the optional gitconfig content for both option group
+	SourceGitconfig = "gitconfig"
```

Comment:
leading dot is specifically allowed in secret data keys:
https://github.com/kubernetes/kubernetes/blob/eb58220e7e3c64d3d6be5bf8741199bd18ecba9f/pkg/api/validation/validation.go#L1540-L1559